### ISSUE TITLE
default query that affects entire library is dangerous

### ISSUE THREAD

**Issue by randybias** [sampsyo/beets#1286]
Maybe this is subjective, but I don't think so... Basically, the default query being '' is extremely dangerous.  A simple typo on the command line can result in something quite bad such as:

beet modify genre="Electro"

Which is the same as:

beet modify genre="Electro" ""

It should take a lot more deliberateness to do a library-wide change like that.  And in fact, something like "beet info" does the right thing in that "beet info" does nothing and you have to issue:

beet info ' '

To get "all".  Which, by the way, is a second argument for making this change.  Probably we want consistency across the various command line arguments as much as possible.

Related, arguably, the query for all should be "*", but maybe that's a whole separate ticket.

**Comment by sampsyo** [comment]
Thanks for bringing this up. I'd like to hear others' opinions about this. Personally, I think `beet ls` should list the entire library, but I would be interested in hearing other stories.

**Comment by randybias** [comment]
I think "beet ls" should list the whole library too.  Maybe that's an exception.  But none of the write operations should default to 'all'.

**Comment by sampsyo** [comment]
It's difficult to see where the line should be drawn. `beet lyrics` should probably fetch all the lyrics, for example, and I'm very accustomed to `beet move` being a global operation by default.

**Comment by randybias** [comment]
Really?  My library will be over 10000 at some point. I don't want lyrics to run as a default on that. Or anything else really. Right now my test set is 300-ish but once I do a full import I will be in the low 1000s. Once I add my non-DJ collection it will be close to 10k.

I can't be the only one with that big a music library. =

**Comment by randybias** [comment]
I feel pretty strongly about this one but I would be happy even if just beet modify didn't have this behavior. There isn't really any reason to run it against all by default. =

**Comment by sampsyo** [comment]
Another option might be to add a confirmation option that confirms when the query arguments are empty. Again, I'd love to hear general opinions on this if anyone's watching the thread.

**Comment by Kraymer** [comment]
I use quasi exclusively the *import* command so this problematic is kinda remote for me but I know that the rare times I used the *beet ls* command, the fact that ``beet ls`` lists the whole library seems counter intuitive.

NB: but I'm a big fan of the `*` character , as made obvious by #1178 where I vouched for it against my better judgment, so ... just my 2c

**Comment by Kraymer** [comment]
After fiddling a bit with *ls* and *rm* command, I have no opinion which is better : * or nothing to consider all files.

> My library will be over 10000 at some point. I don't want lyrics to run as a default on that. 

Why not? beets is good with mass editions that require no human interaction (running in the background), by default it won’t fetch lyrics if the files already have ones, so running it on a whole collection seems a valid use case.

> Another option might be to add a confirmation option that confirms when the query arguments are empty. 

Seeing a long listing with the message "Really modify, move and write tags (Y/n)" was clear enough for  me. But do all the commands display such a confirmation msg before operating ?

I have one suggestion : the *"Modifying X items."* message is not that useful in console mode because as soon as the following listing is longer than terminal height, one have to scroll back to find it.
Why not, put it at the end (eventually with total nb of items in library) like : 

    beet modify genre="Electro"
    [listing]
    1000 of 1000 items. 
    Really modify, move and write tags (Y/n)?

> Probably we want consistency across the various command line arguments as much as possible.

I Agree !

> I would be happy even if just beet modify didn't have this behavior.

consistency ? :wink:

**Comment by randybias** [comment]
Yes. My first choice is consistency across the commands with no default query. I really do think that's the right thing to do here. However I hear sampsyo's preference. In which case my second choice is to make modify less dangerous which is the problem at the heart of this.

If you run a summarize command on a big library you will just get "varies". In other words there is really no time that you wouldn't use modify without a query. It would be very easy for a newbie to unwittingly change a tag for the entire library. As I did. Putting things back is fairly painful BTW. 

So if removing the default query behavior on modify makes sense then it also makes sense to make that consistent across the commands. 

It's a nuanced situation. For non-dangerous commands the default behavior right now is great. For dangerous commands it's the opposite. I would rather err on the side of library integrity rather than convenience but that's just my perspective. 

Re: lyrics maybe that is a bad example. I don't use the command. I know beets is good with massive changes and I am pretty impressed. And hopefully it's good when I hit ^C to abort something I fat fingered. But bad things can clearly happen easily as it stands now. It would be nice if it were harder to blow your library up.

**Comment by sampsyo** [comment]
Warnings like *Modifying N items.* would indeed go a long way to making dangerous operations safe, both with empty queries and when queries are simply more broad than expected.

Maybe we should build a more general "preview operation" functionality, along with configuration to enable and disable it for most/all commands.

**Comment by tlc** [comment]
> Probably we want consistency across the various command line arguments as much as possible.

#679 :)

**Comment by tlc** [comment]
Slightly related -- none of the usage messages mention [QUERY].
So even after checking usage, someone might not realize they probably want one.

$ beet modify -h
Usage: beet modify [options]

Options:
  -h, --help            show this help message and exit
  -M, --nomove          don't move files in library
  -w, --write           write new metadata to files' tags (default)
  -W, --nowrite         don't write metadata (opposite of -w)
  -a, --album           modify whole albums instead of tracks
  -y, --yes             skip confirmation
  -f FORMAT, --format=FORMAT
                        print with custom format

**Comment by sampsyo** [comment]
Yes, this is an unfortunate artifact our hacky use of subcommands in optparse. This would be fixed by a move to Click, as proposed in #1271.

**Comment by alphapapa** [comment]
I was excited to start using Beets, but seeing this issue has made me reconsider.  Potentially destructive operations should never default to operating on all files.  And empty query strings should never default to matching all files.

The `*` character is commonly used to match all terms or files in both command-line utilities and other contexts like search engines.  It should definitely be used.

And potentially destructive operations (including modifying tags in files' internal metadata) should probably default to asking for confirmation before operating on multiple files.

Even in the case of only modifying tags in the database, operating on multiple files without confirmation should probably default to requiring confirmation, because the potential for damage is high.  Even if you could reimport your files to restore metadata to the database, that's a time-consuming, laborious process.  And if files' internal metadata were affected, that would require either fixing and then exporting metadata for all files, or restoring from a backup.  Both of these options are obviously not something anyone would want to have to do.

Combining this with the issue mentioned by tlc of the query string not even being documented makes using Beets downright dangerous if used to modify files in-place.  Accidentally hit Return instead of `'` (the keys are next to each other)?  Well, there goes all your metadata...

**Comment by guibog** [comment]
> I'd love to hear general opinions on this

I think that bulk modifications on files and library data should need a confirmation. 

I am a user of unison and I think they do file synchronization it the right way:
- by default, each change is first reviewed and either confirmad or skipped by the user, and when all changes are reviewed the user confirms the bulk modification
- a --batch flag allow to auto confirm and modify non-conflicting changes
- even when in batch mode, a confirmbigdel option (on by default) will make it safer

Applying this to beets is not obvious, but I think some beets user are very picky on what gets written to files and to the library, and not very prone to let anything be done without their explicit consent. That's why, for example for me, I use beet in no-write no-move mode, because my music is already well organised, and I fear it will mess things up.

**Comment by alphapapa** [comment]
That's a great point about Unison.  I'm also using it in a cron job to sync changes to my music library between computers, and it works very well.  I never have to worry about conflicts, because if there are any, I can resolve them manually by running the command myself whenever I get around to it.  I think Beets should work on the same principle by default.
